Precise macro value type

dingxiangfei2009 opened the pull request:
Fix #923.
Fix #3200. This is due to a fallout from how `libclang` type would be processed with this patch.

This patch introduce a new generator flag `--macro-const-use-ctypes`, which resolves the macro definition expression to a precise type advised by Clang if activated.

A macro expansion result adapter is introduced to maintain compatibility with `cexpr`, which has a limited capacity to resolve to a precise C type.

There is a change to the testing code because up to this point our tests, which can be executed concurrently, use the same path to the precomputed headers. Since this patch introduces a second test that relies on `libclang` macro fallback, the precompiled headers from the two different tests would overwrite each other. The new addition is that the backing files are now temporarily stored at different directories named after the header files under tests.

dingxiangfei2009 commented:
Apparently the clang fallback has race condition. If I set thread number to 1, we don't have truncated output anymore.

dingxiangfei2009 commented:
I have a theory about the test setup leading to the racing. The test setup uses the fallback translation unit backed by a file, which is located in `./.macro_eval.c` by default. If the tests run concurrently, they are racing on this shared file. That probably explains why I get evaluation failure "randomly."

emilio commented on bindgen/options/mod.rs:
```diff
@@ -1109,6 +1109,21 @@ options! {
+            /// Use C platform-specific types for generated macro constants.
+            ///
+            /// This implies `clang_macro_fallback` and takes precedence over
+            /// `default_macro_constant_type`.
```
Would be good to have this be a little less weird, but for now it seems fine.

emilio commented on bindgen/clang.rs:
```diff
@@ -2370,6 +2371,29 @@ impl EvalResult {
+        // Safety: we are only copying the content, not assuming a borrow.
+        // TODO(@dingxiangfei2009): LLVM Libclang does not return the true size
+        // of a string literal, which could be truncated due to a null character
+        // '\0' in the middle.
```
Yeah, hopefully this is rare, but maybe file an upstream issue and reference it here?

dingxiangfei2009 replied:
Oh this is an very old issue: llvm/llvm-project#69749

And apparently a fix: llvm/llvm-project#134551

emilio reviewed (approved):
Looks good, some (optional) nits.

dingxiangfei2009 commented on bindgen/ir/context.rs:
```diff
@@ -2055,10 +2055,12 @@ If you encounter an error missing from this list, please file an issue or a PR!"
             let mut header_names_to_compile = Vec::new();
             let mut header_paths = Vec::new();
             let mut header_includes = Vec::new();
-            let single_header = self.options().input_headers.last().cloned()?;
```
I will break this out as a short PR.

dingxiangfei2009 committed:
Allow macro definitions to pick up precise C types

Upon activating the `--macro-const-use-ctypes` flag, `bindgen` should
generate bindings with precise type resolution to the exact C types.

This flag implies `--clang-macro-fallback` as its pre-requisite.

Signed-off-by: Xiangfei Ding <dingxiangfei2009@protonmail.ch>

dingxiangfei2009 reviewed (commented):
Are there any outstanding concerns about this PR?

dingxiangfei2009 commented:
@emilio Shall we have a look at it again?

dingxiangfei2009 commented:
@emilio any outstanding issues?

dingxiangfei2009 commented:
@emilio what is your opinion on the PR? Any outstanding concerns?
